MekHQ/src/mekhq/campaign/unit/Unit.java

```diff
@@ -1569,20 +1611,23 @@ public double calcInfantryBayWeight(Entity unit) {
      * @param u The unit that we wish to unload from this transport
      */
     public void unloadFromTransportShip(Unit u) {
-        int unitType = u.getEntity().getUnitType();
-        double unitWeight;
-        if (u.getEntity().getUnitType() == UnitType.INFANTRY) {
-            unitWeight = calcInfantryBayWeight(u.getEntity());
-        } else {
-            unitWeight = u.getEntity().getWeight();
-        }
-        for (UUID id : u.getTransportShipId().keySet()) {
-            int bayNumber = u.getTransportShipId().get(id);
-            updateBayCapacity(unitType, unitWeight, true, bayNumber);
-        }
-        removeTransportedUnit(u.getId());
-        u.getTransportShipId().clear();
+        // Only unload and update the transport assignment for a unit
+        // if we have actually transported them.
+        if (removeTransportedUnit(u)) {
```

Comment:
What about the case whereby a unit is assigned to this transport according to the unit but not according to the transport ship?